File: apps/web/src/lib/favourites.ts
/**
 * Favourites (§312, §436; `getting-started` p.34; `ontology.md` §3).
 *
 *     "You can add and remove favorites with the star icon while navigating
 *      the folder structure or **from within an open resource** in a Palantir
 *      platform application." (p.34)
 *
 *     "When you navigate to an individual object view, you can select the star
 *      next to its title to save it as a favorite. This will add the object to
 *      your sidebar… Think of favorites as shortcuts that you can add and
 *      remove to keep frequently used resources close at hand." (p.34)
 *
 * **p.34's sidebar is two asides here, and the reason is the same one §312
 * gave for there being one.** Foundry's sidebar is platform-wide and ours is
 * not, so a favourite is listed where it can be *opened*: an object shortcut
 * beside Saved searches in the Explorer — one a shortcut back to a question,
 * the other back to an answer — and a resource shortcut in the project's
 * browser, which is the screen that opens resources. One store, one cap, two
 * lists, each next to what it leads back to.
 *
 * The server owns every refusal — who may star, how many, whether the type is
 * in this workspace. What lives here is what to *offer*, the division
 * `explorer.ts`, `tags.ts` and `sql-scratchpad.ts` already take.
 */
import type { Favourite } from "./types";

/**
 * What the star says it will do.
 *
 * **What pressing it does, not what the state is.** A control labelled
 * "Favourite" leaves somebody working out whether that is a description or an
 * instruction; `tags.ts` made the same choice about its delete button.
 */
export function starLabel(isFavourite: boolean): string {
  return isFavourite ? "Remove from favourites" : "Add to favourites";
}

/** The glyph, filled once it is kept. */
export function starGlyph(isFavourite: boolean): string {
  return isFavourite ? "★" : "☆";
}

/**
 * What a row in the list is called.
 *
 * The label stored when it was starred, which may be out of date — db 0074
 * takes that trade deliberately, because resolving ten titles is ten reads
 * against the instance store before a sidebar of ten shortcuts can draw.
 *
 * **The fallback is the type's name, never the instance's UUID.** A shortcut
 * reading `aabbccdd-…` is one nobody can choose between; one reading "Ship"
 * at least narrows it, and the type name is already on the row.
 */
export function rowLabel(favourite: Favourite): string {
  const said = favourite.label.trim();
  // **The kind, never the id.** A shortcut reading `aabbccdd-…` is one nobody
  // can choose between; "Dataset" at least narrows it, and since §436 a row
  // may be a resource, whose kind is the only other thing on it.
  return said || favourite.object_type_name || kindWord(favourite) || "Shortcut";
}

/** The kind of resource a shortcut points at, in a word somebody reads —
 *  `code_repo` is a column value, not a noun. */
export function kindWord(favourite: Favourite): string | null {
  const kind = favourite.resource_kind;
  if (!kind) return null;
  return kind === "code_repo"
    ? "Repository"
    : kind.charAt(0).toUpperCase() + kind.slice(1).replace(/_/g, " ");
}

/** The object shortcuts, which the Explorer's aside can open. */
export function objectsOnly(favourites: readonly Favourite[]): Favourite[] {
  return favourites.filter((f) => f.instance_id !== null);
}

/** The resource shortcuts, which the project's browser can open. */
export function resourcesOnly(favourites: readonly Favourite[]): Favourite[] {
  return favourites.filter((f) => f.resource_id !== null);
}

/** Where a resource shortcut goes. The stable id, which is the whole reason
 *  resource ids exist: a link built from slugs stops working the moment
 *  somebody renames either, and §435 made renaming a thing people do. */
export function resourceHref(favourite: Favourite): string {
  return `/r/${favourite.resource_id}`;
}

/** What the resource list says when it is empty. Its own sentence, because it
 *  names a different star in a different place. */
export function resourcesEmptyReason(): string {
  return (
    "None yet. Open a resource and use the star in its header to keep a "
    + "shortcut to it here."
  );
}

/**
 * The line under the name.
 *
 * The type, so two shortcuts with the same label are still tellable apart —
 * and suppressed when the label *is* the type name, because a row that says
 * "Ship / Ship" has spent two lines saying one thing.
 */
export function rowSubtitle(favourite: Favourite): string | null {
  const name = favourite.object_type_name ?? kindWord(favourite);
  if (!name) return null;
  return rowLabel(favourite) === name ? null : name;
}

/**
 * What the empty list says.
 *
 * It names the verb and where to find it, because a star on an object view is
 * not something somebody finds by looking at an empty panel.
 */
export function emptyReason(): string {
  return (
    "None yet. Open an object and use the star beside its title to keep a " +
    "shortcut to it here."
  );
}

/**
 * Whether the star should be drawn at all.
 *
 * **Not while the answer is unknown.** An unfilled star means "not a
 * favourite", so drawing one before the server has said would tell somebody
 * their shortcut is gone — and the press that follows would remove a
 * favourite they still had. Absent for the moment it takes to find out is the
 * honest state, and it is the same rule `status-bar.ts` applies to a problem
 * count nobody has asked for yet.
 */
export function canShowStar(known: boolean): boolean {
  return known;
}

File: apps/api/src/services/favourites.py
"""Favourites: a shortcut to an object or a resource (§312, §436; db 0074, 0100).

    "You can add and remove favorites with the star icon while navigating the
     folder structure or **from within an open resource** in a Palantir
     platform application." (`getting-started` p.34)


    "When you navigate to an individual object view, you can select the star
     next to its title to save it as a favorite. This will add the object to
     your sidebar… Think of favorites as shortcuts that you can add and remove
     to keep frequently used resources close at hand." (p.34)

**Two subjects, one table, one cap** (§436). p.34's first sentence is about
resources and its second is about objects; §312 built the second. The only
difference between the two is the shape of the way back — a resource id, or a
type and an instance — so a second table would have been two services, two
caps and two listings for one idea (§292). The cap is shared because it is a
rule about a *sidebar*: counting the kinds separately would let somebody keep
two hundred shortcuts in a list that stops being "close at hand" at a fraction
of that.

**Whose favourites they are never travels in a request.** db 0074's policy pins
every read and write to the caller, so there is no parameter here that could be
pointed at somebody else's shortcuts — which is the same division
`scratchpad_queries` arrived at the expensive way in §306, where two service
clauses turned out to be doing nothing the policy was not already doing.
"""
from __future__ import annotations

from typing import Any
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncConnection

from ..lib.db import fetch_all, fetch_one
from ..lib.errors import NotFoundError

#: How many objects one person may star in one workspace.
#:
#: **A refusal rather than an eviction**, which is the opposite of §306's
#: scratchpad history and for the opposite reason. A history is a record of
#: what you did and dropping its oldest entry loses nothing anybody chose; a
#: favourite *is* the choice, so silently dropping the least recent would
#: delete a decision. p.34 calls these "frequently used resources close at
#: hand", and a list long enough to need scrolling has stopped being that.
MAX_FAVOURITES = 100


class TooManyFavourites(Exception):
    """The list is full, and the message says what to do about it."""


async def add_object(
    conn: AsyncConnection,
    *,
    user_id: UUID,
    workspace_id: UUID,
    object_type_id: UUID,
    instance_id: UUID,
    label: str,
) -> dict[str, Any]:
    """Star one object.

    **Starring twice is the same star** (db 0074's UNIQUE), so this is safe to
    call from a button whose state arrived a moment ago. The label is refreshed
    on the way through: the object may have been renamed since, and the moment
    somebody stars it again is the one moment we are holding the current name.
    """
    existing = await fetch_one(
        conn,
        """
        SELECT id FROM favourites
         WHERE object_type_id = :tid AND instance_id = :iid
        """,
        {"tid": str(object_type_id), "iid": str(instance_id)},
    )
    if existing is None:
        # Counted only when a row is about to be added, so re-starring
        # something already in a full list is not refused for being full.
        await _room_for_one_more(conn, workspace_id=workspace_id)

    row = await fetch_one(
        conn,
        """
        INSERT INTO favourites
               (user_id, workspace_id, object_type_id, instance_id, label)
        VALUES (:uid, :wid, :tid, :iid, :label)
        ON CONFLICT (user_id, object_type_id, instance_id) DO UPDATE
           SET label = EXCLUDED.label
        RETURNING id, object_type_id, instance_id, label, created_at
        """,
        {"uid": str(user_id), "wid": str(workspace_id),
         "tid": str(object_type_id), "iid": str(instance_id), "label": label[:500]},
    )
    assert row is not None
    return dict(row)


async def remove_object(
    conn: AsyncConnection, *, object_type_id: UUID, instance_id: UUID
) -> None:
    """Unstar one object.

    Addressed by *what it points at* rather than by the favourite's own id,
    because that is what the caller has: the star sits on an object view, which
    knows the object and has never been told the row's id.
    """
    row = await fetch_one(
        conn,
        """
        DELETE FROM favourites
         WHERE object_type_id = :tid AND instance_id = :iid
        RETURNING id
        """,
        {"tid": str(object_type_id), "iid": str(instance_id)},
    )
    if row is None:
        raise NotFoundError("this favourite")


async def listing(
    conn: AsyncConnection, *, workspace_id: UUID
) -> list[dict[str, Any]]:
    """This person's favourites in this workspace, newest first.

    **Newest first, not alphabetical.** p.34's shortcuts are "frequently used
    resources close at hand", and what somebody starred most recently is what
    they are working on now. A list sorted by a label the reader did not choose
    would bury today's work under a name beginning with A.

    The type's display name is joined so a row can say what kind of thing it
    is; the object's own label is stored (db 0074) so ten shortcuts do not cost
    ten reads against the instance store.
    """
    rows = await fetch_all(
        conn,
        """
        SELECT f.id, f.object_type_id, f.instance_id, f.resource_id, f.label,
               f.created_at,
               t.display_name AS object_type_name,
               r.kind AS resource_kind
          -- **LEFT JOINs, because a row has exactly one subject** (db 0100).
          -- An inner join to either table would silently drop the other kind,
          -- which is the sidebar quietly losing half of itself.
          FROM favourites f
          LEFT JOIN object_types t ON t.id = f.object_type_id
          LEFT JOIN resources r ON r.id = f.resource_id
         WHERE f.workspace_id = :wid
         ORDER BY f.created_at DESC
        """,
        {"wid": str(workspace_id)},
    )
    return [dict(r) for r in rows]


async def starred(
    conn: AsyncConnection, *, object_type_id: UUID, instance_id: UUID
) -> bool:
    """Whether this object is one of the caller's favourites.

    Its own read rather than a scan of `listing`, because the star is drawn on
    an object view that has no reason to have fetched the whole list — and a
    view that fetched a hundred favourites to decide the state of one button
    would be paying the sidebar's cost on a screen that has no sidebar.
    """
    row = await fetch_one(
        conn,
        """
        SELECT 1 AS yes FROM favourites
         WHERE object_type_id = :tid AND instance_id = :iid
        """,
        {"tid": str(object_type_id), "iid": str(instance_id)},
    )
    return row is not None


# ---- the other subject (§436; db 0100) --------------------------------------
async def _room_for_one_more(
    conn: AsyncConnection, *, workspace_id: UUID
) -> None:
    """Refuse when the sidebar is full.

    Shared by both kinds, and counted across both: the cap is a rule about how
    long a list of shortcuts may be before it stops being one.
    """
    count = await fetch_one(
        conn,
        "SELECT count(*) AS n FROM favourites WHERE workspace_id = :wid",
        {"wid": str(workspace_id)},
    )
    assert count is not None
    if int(count["n"]) >= MAX_FAVOURITES:
        raise TooManyFavourites(
            f"You have {MAX_FAVOURITES} favourites in this workspace, which "
            "is as many as this list holds. Remove one you have finished "
            "with — favourites are shortcuts to what you are working on now."
        )


async def add_resource(
    conn: AsyncConnection,
    *,
    user_id: UUID,
    workspace_id: UUID,
    resource_id: UUID,
    label: str,
) -> dict[str, Any]:
    """Star one resource — p.34's "from within an open resource".

    The label is refreshed on the way through for the reason the object half
    does it: the resource may have been renamed since (§435 made that possible
    from the screen), and the moment somebody stars it again is the one moment
    we are holding the current name.
    """
    existing = await fetch_one(
        conn,
        "SELECT id FROM favourites WHERE resource_id = :rid",
        {"rid": str(resource_id)},
    )
    if existing is None:
        await _room_for_one_more(conn, workspace_id=workspace_id)

    row = await fetch_one(
        conn,
        """
        INSERT INTO favourites (user_id, workspace_id, resource_id, label)
        VALUES (:uid, :wid, :rid, :label)
        ON CONFLICT (user_id, resource_id) DO UPDATE
           SET label = EXCLUDED.label
        RETURNING id, resource_id, label, created_at
        """,
        {"uid": str(user_id), "wid": str(workspace_id),
         "rid": str(resource_id), "label": label[:500]},
    )
    assert row is not None
    return dict(row)


async def remove_resource(conn: AsyncConnection, *, resource_id: UUID) -> None:
    """Unstar one resource, by what it points at rather than by the row's id —
    the star sits on an application's header, which knows the resource and has
    never been told the favourite's id."""
    row = await fetch_one(
        conn,
        "DELETE FROM favourites WHERE resource_id = :rid RETURNING id",
        {"rid": str(resource_id)},
    )
    if row is None:
        raise NotFoundError("this favourite")


async def resource_starred(conn: AsyncConnection, *, resource_id: UUID) -> bool:
    """Whether this resource is one of the caller's favourites."""
    row = await fetch_one(
        conn,
        "SELECT 1 AS yes FROM favourites WHERE resource_id = :rid",
        {"rid": str(resource_id)},
    )
    return row is not None
